runtime/jcl/common/criusupport.c

```diff
@@ -44,3 +54,94 @@ Java_com_ibm_jvm_CRIUSupport_isCRIUSupportEnabledImpl(JNIEnv *env, jclass unused
 
 	return res;
 }
+
+jint JNICALL
+Java_com_ibm_jvm_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
+							jclass unused,
+							jobject checkPointDir,
+							jboolean keepRunning,
+							jboolean shellJob,
+							jboolean extUnixSupport,
+							jint logLevel,
+							jobject logFile)
+{
+	jint rc = UNSUPPORTED_OPERATION;
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	char directoryBuf[STRING_BUFFER_SIZE] = {0};
+	char *directoryUTF = NULL;
+	char logFileBuf[STRING_BUFFER_SIZE] = {0};
+	char *logFileUTF = NULL;
+
+	if (J9_ARE_ALL_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_ENABLE_CRIU_SUPPORT)) {
+#if defined(LINUX)
+		j9object_t cpDir = NULL;
+		j9object_t log = NULL;
+		IDATA dirFD = 0;
+		PORT_ACCESS_FROM_VMC(currentThread);
+
+		vmFuncs->internalEnterVMFromJNI(currentThread);
+
+		cpDir = J9_JNI_UNWRAP_REFERENCE(checkPointDir);
+		log = J9_JNI_UNWRAP_REFERENCE(logFile);
+
+		directoryUTF = vmFuncs->copyStringToUTF8WithMemAlloc(currentThread, cpDir, J9_STR_NULL_TERMINATE_RESULT, "", 0, directoryBuf, STRING_BUFFER_SIZE, NULL);
+		if (NULL == directoryUTF) {
+			vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto release;
+		}
+
+		logFileUTF = vmFuncs->copyStringToUTF8WithMemAlloc(currentThread, log, J9_STR_NULL_TERMINATE_RESULT, "", 0, logFileBuf, STRING_BUFFER_SIZE, NULL);
+		if (NULL == logFileUTF) {
+			vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto freeDir;
+		}
+
+		dirFD = open(directoryUTF, O_DIRECTORY);
```

Comment:
What if the current encoding isn't UTF8? The filename should be converted to the current encoding in Java.

Reply:
Or omrstr_convert() can be used.

Reply:
Who is responsible for closing `dirFD`?

Reply:
The encoding issue isn't resolved.

Reply:
the string is converted to the `StandardCharsets.UTF_8` charset on the Java side

Reply:
This only works because (a) you are probably testing in English and (b) your LANG is likely set to `en_US.UTF-8`. If you set something else, such as `en_US.iso88591` then UTF8 isn't going to work. You need to convert to the current platform charset, not always UTF8. Either pass in the converted bytes to the native, or see about using omrstr_convert() to convert the UTF-16 in the String to the platform encoding. `new String(cpDataDir.getBytes(), StandardCharsets.UTF_8)` is just going to corrupt your data. It's converting to bytes in the current encoding, and then converting those bytes back to UTF-16 assuming they (and the current encoding) are UTF8. If they are not UTF8 you'll get truncated or garbage results.

Reply:
You'll get this: `This method always replaces malformed-input and unmappable-character sequences with this charset's default replacement string`

Reply:
Is it not as simple as replacing `new String(cpDataDir.getBytes(), StandardCharsets.UTF_8)` with `cpDataDir.getBytes(StandardCharsets.UTF_8)`?

Reply:
Closer, except StandardCharsets.UTF_8 needs to be the current detected encoding, such as in the system property `file.encoding`. I think just using `getBytes()` is correct, and passing those bytes to the native should fix it. Except for Windows, which should use Unicode, but that platform isn't supported for this. If Windows is a future possibility keep passing the String and use omrstr_convert( ) to convert to the platform encoding.

Reply:
Actually Windows will still work with passing bytes. Basically on Windows you can convert to UTF8 bytes, and then in the native convert the UTF8 back to Unicode and call the Unicode functions.